node/services.js

```diff
@@ -0,0 +1,129 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var errors = require('./errors');
+
+function TChannelServices() {
+    var self = this;
+    self.services = Object.create(null);
+    self.maxPendingForService = Infinity;
+    self.maxPending = Infinity;
+    self.pending = 0;
+}
+
+TChannelServices.prototype.considerDecline = function considerDecline(req) {
+    var self = this;
+    if (self.pending >= self.maxPending) {
+        return errors.MaxPendingError({
+            pending: self.pending
+        });
+    }
+    if (!req.options.serviceName) {
+        return;
+    }
+    var key = mangle(req.options.serviceName);
+    var service = self.services[key];
+    return service && service.considerDecline();
+};
+
+TChannelServices.prototype.onRequest = function onRequest(req) {
+    var self = this;
+    self.pending++;
+    if (!req.options.serviceName) {
+        return;
+    }
+    var key = mangle(req.options.serviceName);
+    var service = self.services[key];
+    if (!service) {
+        service = new TChannelService();
+        service.serviceName = req.options.serviceName;
+        service.maxPending = self.maxPendingForService;
+        self.services[key] = service;
+    }
+    service.onRequest();
+};
+
+TChannelServices.prototype.onRequestResponse = function onRequestResponse(req) {
+    var self = this;
+    self.pending--;
+    if (!req.options.serviceName) {
+        return;
+    }
+    var key = mangle(req.options.serviceName);
+    var service = self.services[key];
+    service.onRequestResponse();
+};
+
+TChannelServices.prototype.onRequestError = function onRequestError(req) {
+    var self = this;
+    self.pending--;
+    if (!req.options.serviceName) {
+        return;
+    }
+    var key = mangle(req.options.serviceName);
+    var service = self.services[key];
+    service.onRequestError();
+};
+
+function TChannelService() {
+    var self = this;
+    self.serviceName = null;
+    self.maxPending = Infinity;
+    self.pending = 0;
+}
+
+TChannelService.prototype.considerDecline = function considerDecline() {
+    var self = this;
+    if (self.pending >= self.maxPending) {
+        return errors.MaxPendingForServiceError({
+            serviceName: self.serviceName,
+            pending: self.pending
+        });
+    }
+};
+
+TChannelService.prototype.onRequest = function onRequest() {
+    var self = this;
+    self.pending += 1;
+};
+
+TChannelService.prototype.onRequestResponse = function onRequestResponse() {
+    var self = this;
+    self.pending -= 1;
+};
+
+TChannelService.prototype.onRequestError = function onRequestError() {
+    var self = this;
+    self.pending -= 1;
+};
+
+// Mangling service names ensures that no joker can confuse us with
+// '__proto__'.
+function mangle(serviceName) {
```

Comment:
suggest name trunderdunder.... dang this is ugly, but I guess the lastIndexOf trick is really a clever way of keeping the overhead minimal, so let the flame graph will out.